### PR TITLE
FermiNet changes for weight initialization and minor type fix

### PR DESCRIPTION
**Pull request by shaipranesh2**
## Description

Fix #(issue)

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/ferminet.py [hunk 1]
```diff
@@ -139,7 +139,8 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:
             (one_electron_vector, one_electron_distance.unsqueeze(-1)), dim=3)
         one_electron = torch.reshape(one_electron.permute(0, 2, 1, 3),
                                      (self.batch_size, self.total_electron, -1))
-        one_electron_vector_permuted = one_electron_vector.permute(0, 2, 1, 3)
+        one_electron_vector_permuted = one_electron_vector.permute(0, 2, 1,
+                                                                   3).float()
```
Comment (shaipranesh2) [inline deepchem/models/torch_models/ferminet.py:143]: This is the minor type fix as ferminet layers requires its inputs to be in float data type, but the the one_electron_vector_permuted is a double datatype (from torch.permute)

File: deepchem/models/torch_models/layers.py [hunk 2]
```diff
@@ -5556,19 +5551,20 @@ def forward(self, one_electron: torch.Tensor, two_electron: torch.Tensor):
...
-                    two_electron_tmp[:, i, :, :] = torch.tanh(self.w[l](
-                        two_electron[:, i, :, :])) + two_electron[:, i, :]
-            one_electron = one_electron_tmp
-            two_electron = two_electron_tmp
+                    one_electron_tmp.append(
+                        (torch.tanh(self.v[l](f)) + one_electron[:, i, :]))
+                    two_electron_tmp.append(
+                        (torch.tanh(self.w[l](two_electron[:, i, :, :])) +
+                         two_electron[:, i, :, :]))
+            one_electron = torch.stack(one_electron_tmp, dim=1)
```
Comment (shaipranesh2) [inline deepchem/models/torch_models/layers.py:5566]: changes to accommodate for single electron ions in the next PR

File: deepchem/models/torch_models/layers.py [hunk 3]
```diff
@@ -5674,43 +5669,49 @@ def forward(self, one_electron: torch.Tensor,
...
-        return psi, psi_up, psi_down
+                    det.append(((torch.sum(
+                        (self.envelope_w[one_d_index] * one_electron[:, j, :]) +
+                        self.envelope_g[one_d_index],
+                        dim=1)) * torch.sum(torch.exp(-torch.abs(
+                            torch.norm(one_electron_vector_permuted[:, j, :, :]
+                                       @ self.sigma[one_d_index],
+                                       dim=2))) * self.pi[one_d_index].T,
+                                            dim=1)))
+            psi_down.append(
```
Comment (shaipranesh2) [inline deepchem/models/torch_models/layers.py:5704]: changes to accommodate for single electron ions in the next PR

File: deepchem/models/torch_models/layers.py [hunk 4]
```diff
@@ -5556,19 +5551,20 @@ def forward(self, one_electron: torch.Tensor, two_electron: torch.Tensor):
                                             dim=1)
                 if l == 0 or (self.n_one[l] != self.n_one[l - 1]) or (
                         self.n_two[l] != self.n_two[l - 1]):
-                    one_electron_tmp[:, i, :] = torch.tanh(
-                        self.v[l](f)) + self.projection_module[0](
-                            one_electron[:, i, :])
-                    two_electron_tmp[:, i, :, :] = torch.tanh(self.w[l](
-                        two_electron[:, i, :, :])) + self.projection_module[1](
-                            two_electron[:, i, :, :])
+                    one_electron_tmp.append((torch.tanh(self.v[l](f))) +
```
Comment (shaipranesh2) [inline deepchem/models/torch_models/layers.py:5554]: earlier version was not working for single electron ions, refactored the code accordingly to make changes to work for all systems

File: deepchem/models/torch_models/layers.py [hunk 5]
```diff
@@ -5674,43 +5669,49 @@ def forward(self, one_electron: torch.Tensor,
...
+                            torch.norm(one_electron_vector_permuted[:, j, :, :]
+                                       @ self.sigma[one_d_index],
+                                       dim=2))) * self.pi[one_d_index].T,
+                                            dim=1)))
+            psi_down.append(
+                torch.reshape(torch.stack(det, dim=1),
+                              (self.batch_size, self.spin[1], self.spin[1])))
+
+            d_down = torch.det(psi_down[-1])
+            d_up = torch.det(psi_up[-1])
```
Comment (shaipranesh2) [inline deepchem/models/torch_models/layers.py:5709]: earlier version was not working for single electron ions, refactored the code accordingly to make changes to work for all systems

File: deepchem/models/torch_models/layers.py [hunk 6]
```diff
@@ -5674,43 +5669,49 @@ def forward(self, one_electron: torch.Tensor,
...
-            d_down = torch.det(psi_down[:, k, :, :].clone())
-            d_up = torch.det(psi_up[:, k, :, :].clone())
-            det = d_up * d_down
-            psi = psi + det
-        return psi, psi_up, psi_down
+                    det.append(((torch.sum(
+                        (self.envelope_w[one_d_index] * one_electron[:, j, :]) +
+                        self.envelope_g[one_d_index],
+                        dim=1)) * torch.sum(torch.exp(-torch.abs(
+                            torch.norm(one_electron_vector_permuted[:, j, :, :]
```
Comment (shaipranesh2) [inline deepchem/models/torch_models/layers.py:5700]: earlier version was not working for single electron ions, refactored the code accordingly to make changes to work for all systems

File: deepchem/models/torch_models/layers.py (hunk 4 above)
Comment (rbharath) [reply deepchem/models/torch_models/layers.py:5554]: Can you add a unit test in a follow up PR for single electron ions?

**Review by rbharath — APPROVED** [review]
LGTM

**Pull request merged**